test/test_connectionpool.py

```diff
@@ -226,6 +226,45 @@ def test_contextmanager(self):
 
         self.assertRaises(Empty, old_pool_queue.get, block=False)
 
+    def test_connection_recycling(self):
+        '''
+        Test that connections are recycled to the pool on 
+        connection/readtimeout errors where no http response 
+        is received.
+        '''
+        poolsize = 2
+        with HTTPConnectionPool('google.com', maxsize=poolsize, 
+                                  retries=1, block=True) as pool:
+
+            self.assertEqual(pool.pool.qsize(), poolsize)
+
+            # set a silly read timeout to force timeout exception. 
+            # We won't get a response for this and so the conn won't 
+            # be implicitly returned to the pool.
+            pool.timeout._read = 0
+            self.assertRaises(MaxRetryError, pool.urlopen, 'GET', 
```

Comment:
Does this make an external request to actual google.com? We don't allow external requests in our test suite to keep it fast and runnable offline.

Could either mock the connection objects or use our with_dummyserver suite.

On a related note, would be nice if we had somekind of FakeConnection object we could use in cases like this. (Just thinking aloud.)